src/main/java/org/tarantool/TarantoolClientImpl.java

```diff
@@ -66,14 +65,30 @@
     protected Thread reader;
     protected Thread writer;
 
+    /**
+     * The condition variable to signal a reconnection is needed from reader /
+     * writer threads and waiting for that signal from the reconnection thread.
+     *
+     * The lock variable to access this condition.
+     *
+     * XXX: Maybe it is better to move signalling / waiting support to the
+     * state helper (or fsm when we'll reimplement it in that way). Maybe
+     * additional RECONNECTION_NEEDED state is required to distinguish whether
+     * we're in the reconnection state already or just want to reconnect (don't
+     * sure).
+     *
+     * @see #awaitReconnection()
+     * @see #trySignalForReconnection()
+     */
+    protected final ReentrantLock connectorLock = new ReentrantLock();
+    protected final Condition reconnectRequired = connectorLock.newCondition();
```

Comment:
I think it is responsibility of the state helper to implement appropriate locking / signaling and provide this functionality for the connector. See also XXX comment above.